core/trino-main/src/main/java/io/trino/exchange/DirectExchangeInput.java

```diff
@@ -51,6 +53,25 @@ public String getLocation()
         return location;
     }
 
+    @Override
+    public boolean equals(Object o)
```

Comment:
do you need the equals/hashCode for `ExchangeInput` implementations. Why cannot we keep the identity based equality? Is it possible that `updateOutputLocations` would be called with `ExchangeInput` which was already registered previously?